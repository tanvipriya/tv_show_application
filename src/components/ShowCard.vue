<template>
  <div class="show-card">
    <img :src="show.image?.medium || placeholder" :alt="show.name" />
    <h3 class="show-name">{{ show.name }}</h3>
    <p class="show-rating">‚≠ê {{ show.rating?.average || "N/A" }}</p>
    <p class="show-year">{{ show.premiered?.slice(0, 4) || "N/A" }}</p>
  </div>
</template>

<script setup>
defineProps({
  show: Object,
});

const placeholder = "https://via.placeholder.com/150x225?text=No+Image";
</script>

<style scoped>
.show-card {
  min-width: 150px;
  margin: 0 8px;
  text-align: center;
  flex-shrink: 0;
}
.show-card img {
  width: 100%;
  border-radius: 8px;
  object-fit: cover;
}
.show-name {
  font-size: 0.9rem;
  margin: 4px 0 2px 0;
}
.show-rating,
.show-year {
  font-size: 0.8rem;
  margin: 1px 0;
}
</style>
