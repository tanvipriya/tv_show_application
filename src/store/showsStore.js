import { defineStore } from 'pinia';
import axios from 'axios';

export const useShowsStore = defineStore('shows', {
    state: () => ({
        showsByGenre: {}, // { drama: [...], comedy: [...] }
        loading: false,
        error: null
    }),
    actions: {
        async fetchShowsByGenre(genre) {
            this.loading = true;
            this.error = null;
            try {
                // Fetch first 2 pages (TVMaze has 250 shows per page)
                const pages = [0, 1];
                let allShows = [];

                for (const page of pages) {
                    const res = await axios.get(`https://api.tvmaze.com/shows?page=${page}`);
                    allShows = allShows.concat(res.data);
                }

                // Filter by genre
                const filtered = allShows.filter((show) => show.genres.includes(genre));

                // Sort by rating descending (null rating goes last)
                filtered.sort((a, b) => (b.rating.average || 0) - (a.rating.average || 0));

                this.showsByGenre[genre] = filtered;
            } catch (err) {
                this.error = err.message || 'Failed to fetch shows';
            } finally {
                this.loading = false;
            }
        }
    }
});
