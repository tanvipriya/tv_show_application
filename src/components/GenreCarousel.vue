<template>
  <div class="genre-carousel">
    <div class="carousel-header">
      <h2>{{ genre }}</h2>
      <router-link
        :to="{ name: 'GenreView', params: { genre } }"
        class="see-all"
      >
        See All
      </router-link>
    </div>

    <div class="carousel-container">
      <router-link
        v-for="show in shows"
        :key="show.id"
        :to="{ name: 'ShowDetail', params: { id: show.id } }"
        class="show-link"
      >
        <ShowCard :show="show" />
      </router-link>
    </div>

    <p v-if="loading">Loading...</p>
    <p v-if="error" class="error">{{ error }}</p>
  </div>
</template>

<script setup>
import { onMounted, computed } from "vue";
import { useShowsStore } from "@/store/showsStore";
import ShowCard from "./ShowCard.vue";

const props = defineProps({
  genre: String,
});

const store = useShowsStore();

// Compute shows for this genre
const shows = computed(() => store.showsByGenre[props.genre] || []);
const loading = computed(() => store.loading);
const error = computed(() => store.error);

onMounted(async () => {
  // Load first batch if not loaded yet
  if (!store.showsByGenre[props.genre]) {
    await store.fetchNextPage(props.genre); // <--- updated function
  }
});
</script>

<style scoped>
.genre-carousel {
  margin-bottom: 24px;
}

.carousel-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 8px;
}

.carousel-container {
  display: flex;
  overflow-x: auto;
  scroll-behavior: smooth;
  gap: 12px;
  padding-bottom: 8px;
}

.carousel-container::-webkit-scrollbar {
  height: 8px;
}

.carousel-container::-webkit-scrollbar-thumb {
  background: #888;
  border-radius: 4px;
}

.show-link {
  text-decoration: none;
  color: inherit;
}

.see-all {
  font-size: 0.85rem;
  color: #004c4d;
  text-decoration: none;
}

.error {
  color: red;
}
</style>
