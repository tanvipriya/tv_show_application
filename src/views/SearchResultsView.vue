<template>
  <div class="search-view">
    <input
      type="text"
      v-model="query"
      placeholder="Search TV shows..."
      @input="onSearch"
      class="search-input"
    />

    <!-- Loading skeletons -->
    <div v-if="loading" class="results">
      <div class="show-card-skeleton" v-for="n in 6" :key="n"></div>
    </div>

    <!-- Error -->
    <div v-if="error" class="error">{{ error }}</div>

    <!-- Results -->
    <div v-if="results.length" class="results">
      <div
        v-for="show in results"
        :key="show.id"
        class="show-card"
        @click="goToShowDetail(show.id)"
      >
        <ShowCard :show="show" @toggle-favorite="toggleFavorite(show.id)" />
      </div>
    </div>

    <p v-else-if="query && !loading">No results found</p>
  </div>
</template>

<script setup>
import { ref, computed } from "vue";
import { useRouter } from "vue-router";
import { useShowsStore } from "@/store/showsStore";
import ShowCard from "@/components/ShowCard.vue";
import debounce from "lodash/debounce";

const store = useShowsStore();
const router = useRouter();
const query = ref("");

const results = computed(() => {
  if (!query.value) return [];
  const lowerQuery = query.value.toLowerCase();
  if (store.searchResults.length) {
    return store.searchResults.filter((show) =>
      show.name.toLowerCase().includes(lowerQuery)
    );
  }
  const allShows = Object.values(store.showsByGenre).flat();
  return allShows.filter((show) =>
    show.name.toLowerCase().includes(lowerQuery)
  );
});

const loading = computed(() => store.loading);
const error = computed(() => store.error);

function goToShowDetail(showId) {
  router.push({ name: "ShowDetail", params: { id: showId } });
}

function toggleFavorite(showId) {
  store.toggleFavorite(showId);
}

const onSearch = debounce(async () => {
  if (!query.value) return;
  await store.fetchShowsByName(query.value);
}, 300);
</script>

<style scoped>
.search-view {
  padding: 16px;
}
.search-input {
  width: 100%;
  padding: 8px 12px;
  font-size: 1rem;
  margin-bottom: 16px;
}
.results {
  display: flex;
  flex-wrap: wrap;
  gap: 12px;
}
.show-card {
  cursor: pointer;
  width: 150px;
}
.show-card-skeleton {
  width: 150px;
  height: 220px;
  background: linear-gradient(90deg, #eee 25%, #ddd 50%, #eee 75%);
  background-size: 200% 100%;
  animation: shimmer 1.5s infinite;
  border-radius: 8px;
}
@keyframes shimmer {
  0% {
    background-position: -200% 0;
  }
  100% {
    background-position: 200% 0;
  }
}
.error {
  color: red;
  margin-top: 8px;
}
</style>
