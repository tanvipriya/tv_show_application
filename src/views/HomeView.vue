<template>
  <div>
    <h1>TV Shows Dashboard</h1>
    <GenreCarousel genre="Drama" />
    <GenreCarousel genre="Comedy" />
    <GenreCarousel genre="Action" />
    <GenreCarousel genre="Sports" />
    <GenreCarousel genre="Horror" />
  </div>
</template>

<script setup>
import GenreCarousel from "@/components/GenreCarousel.vue";
</script>

<style scoped>
h1 {
  font-size: 1.5rem;
  margin-bottom: 16px;
}
</style>
