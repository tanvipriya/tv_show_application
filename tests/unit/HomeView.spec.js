// tests/unit/HomeView.spec.js
import { mount, flushPromises } from "@vue/test-utils";
import { createTestingPinia } from "@pinia/testing";
import { createRouter, createMemoryHistory } from "vue-router";
import HomeView from "@/views/HomeView.vue";
import GenreCarousel from "@/components/GenreCarousel.vue";
import ShowCard from "@/components/ShowCard.vue";
import SearchResultsView from "@/views/SearchResultsView.vue";
import { useShowsStore } from "@/store/showsStore";

describe("HomeView.vue", () => {
    let router;

    beforeEach(async () => {
        router = createRouter({
            history: createMemoryHistory(),
            routes: [
                { name: "Home", path: "/", component: { template: "<div>Home</div>" } },
                { name: "ShowDetail", path: "/show/:id", component: { template: "<div>Show Detail</div>" } },
            ],
        });

        await router.push("/"); // now this matches
        await router.isReady();
    });


    it("renders correctly and matches snapshot", () => {
        const wrapper = mount(HomeView, {
            global: {
                plugins: [router, createTestingPinia({ stubActions: false })],
                stubs: {
                    GenreCarousel: true,
                    ShowCard: true,
                    SearchResultsView: true,
                },
            },
        });

        // Static content
        expect(wrapper.find(".search-title").text()).toContain("TV Shows Dashboard");
        expect(wrapper.find(".search-subtitle").text()).toBe("Find your favorite shows quickly");

        // Snapshot
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("renders genre carousels when searchQuery is empty", () => {
        const wrapper = mount(HomeView, {
            global: {
                plugins: [router, createTestingPinia()],
                stubs: {
                    GenreCarousel: true,
                    ShowCard: true,
                    SearchResultsView: true,
                },
            },
        });

        expect(wrapper.findAllComponents(GenreCarousel).length).toBe(5);
    });

    it("filters shows when searchQuery is set and shows results", async () => {
        const pinia = createTestingPinia({ stubActions: false });
        const store = useShowsStore();
        store.showsByGenre = {
            Drama: [{ id: 1, name: "Breaking Bad" }],
            Comedy: [{ id: 2, name: "Friends" }],
        };

        const wrapper = mount(HomeView, {
            global: {
                plugins: [router, pinia],
                stubs: {
                    GenreCarousel: true,
                    // ShowCard stub renders the show name for test assertions
                    ShowCard: {
                        props: ["show"],
                        template: `<div class="stub-show-card">{{ show.name }}</div>`,
                    },
                    // SearchResultsView stub with v-model emit
                    SearchResultsView: {
                        template:
                            '<input v-model="modelValue" @input="$emit(\'update:modelValue\', $event.target.value)" />',
                        props: ["modelValue"],
                    },
                },
            },
        });

        // Simulate user typing in SearchResultsView
        const searchInput = wrapper.findComponent(SearchResultsView);
        await searchInput.vm.$emit("update:modelValue", "Breaking");
        await flushPromises();

        // Filtered results should appear
        expect(wrapper.findAllComponents(ShowCard).length).toBe(1);
        expect(wrapper.text()).toContain("Breaking Bad");

        // Clicking show card triggers navigation
        const goToShowDetailSpy = jest.spyOn(wrapper.vm, "goToShowDetail");
        await wrapper.find(".stub-show-card").trigger("click");
        expect(goToShowDetailSpy).toHaveBeenCalledWith(1);
    });

    it("shows 'No results found' if searchQuery has no matches", async () => {
        const pinia = createTestingPinia({ stubActions: false });
        const store = useShowsStore();
        store.showsByGenre = {
            Drama: [{ id: 1, name: "Breaking Bad" }],
        };

        const wrapper = mount(HomeView, {
            global: {
                plugins: [router, pinia],
                stubs: {
                    GenreCarousel: true,
                    ShowCard: {
                        props: ["show"],
                        template: `<div class="stub-show-card">{{ show.name }}</div>`,
                    },
                    SearchResultsView: {
                        template:
                            '<input v-model="modelValue" @input="$emit(\'update:modelValue\', $event.target.value)" />',
                        props: ["modelValue"],
                    },
                },
            },
        });

        const searchInput = wrapper.findComponent(SearchResultsView);
        await searchInput.vm.$emit("update:modelValue", "Friends");
        await flushPromises();

        expect(wrapper.text()).toContain('No results found for "Friends"');
        // Genre carousels should be hidden
        expect(wrapper.findAllComponents(GenreCarousel).length).toBe(0);
    });
});
