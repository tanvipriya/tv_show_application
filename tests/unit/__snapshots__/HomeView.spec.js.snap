// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeView.vue renders correctly and matches snapshot 1`] = `
<div>
  <h1 class="search-title"><i class="fas fa-tv"></i> TV Shows Dashboard</h1>
  <p class="search-subtitle">Find your favorite shows quickly</p><!-- Search input -->
  <search-results-view-stub modelvalue=""></search-results-view-stub><!-- Search results -->
  <!--v-if-->
  <!-- Genre carousels (hide if searchQuery is not empty) -->
  <div>
    <genre-carousel-stub genre="Drama"></genre-carousel-stub>
    <genre-carousel-stub genre="Comedy"></genre-carousel-stub>
    <genre-carousel-stub genre="Action"></genre-carousel-stub>
    <genre-carousel-stub genre="Sports"></genre-carousel-stub>
    <genre-carousel-stub genre="Horror"></genre-carousel-stub>
  </div>
</div>
`;
